include/libtorrent/session_params.hpp

```diff
@@ -0,0 +1,131 @@
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_SESSION_PARAMS_HPP_INCLUDED
+#define TORRENT_SESSION_PARAMS_HPP_INCLUDED
+
+#include <functional>
+#include <memory>
+#include <vector>
+
+#include "libtorrent/config.hpp"
+#include "libtorrent/aux_/export.hpp"
+#include "libtorrent/settings_pack.hpp"
+#include "libtorrent/io_context.hpp"
+#include "libtorrent/kademlia/dht_state.hpp"
+#include "libtorrent/session_types.hpp"
+#include "libtorrent/kademlia/dht_storage.hpp"
+
+#if TORRENT_ABI_VERSION <= 2
+#include "libtorrent/kademlia/dht_settings.hpp"
+#endif
+
+namespace libtorrent {
+
+struct plugin;
+struct disk_interface;
+struct counters;
+
+using disk_io_constructor_type = std::function<std::unique_ptr<disk_interface>(
+	io_context&, settings_interface const&, counters&)>;
+
+TORRENT_VERSION_NAMESPACE_3
+
+// The session_params is a parameters pack for configuring the session
+// before it's started.
+struct TORRENT_EXPORT session_params
+{
+	// This constructor can be used to start with the default plugins
+	// (ut_metadata, ut_pex and smart_ban). The default values in the
+	// settings is to start the default features like upnp, NAT-PMP,
+	// and dht for example.
+	session_params(settings_pack&& sp); // NOLINT
+	session_params(settings_pack const& sp); // NOLINT
+	session_params();
+	~session_params();
+
+	// This constructor helps to configure the set of initial plugins
+	// to be added to the session before it's started.
+	session_params(settings_pack&& sp
+		, std::vector<std::shared_ptr<plugin>> exts);
+	session_params(settings_pack const& sp
+		, std::vector<std::shared_ptr<plugin>> exts);
+
+	// hidden
+	session_params(session_params const&);
+	session_params(session_params&&);
+	session_params& operator=(session_params const&) &;
+	session_params& operator=(session_params&&) &;
+
+	settings_pack settings;
+
+	std::vector<std::shared_ptr<plugin>> extensions;
+
+#if TORRENT_ABI_VERSION <= 2
+#ifdef _MSC_VER
+#pragma warning(push, 1)
+#pragma warning( disable : 4996 ) // warning C4996: X: was declared deprecated
+#endif
+#if defined __GNUC__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#endif
+	// this is deprecated. Use the dht_* settings instead.
+	dht::dht_settings dht_settings;
+#if defined __GNUC__
+#pragma GCC diagnostic pop
+#endif
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif
+
+#endif
+
+	dht::dht_state dht_state;
+
+	dht::dht_storage_constructor_type dht_storage_constructor;
+
+	disk_io_constructor_type disk_io_constructor;
+
+	// this container can be used by extensions/plugins to store settings. It's
+	// primarily here to make it convenient to save and restore state across
+	// sessions, using read_session_params() and write_session_params().
+	std::map<std::string, std::string> ext_state;
```

Comment:
👍

Reply:
heh. technically, this part belongs in a future patch